test/unit/bindingpatterns.spec.ts

```diff
@@ -82,3 +83,103 @@ test.each([
     );
     expect(result).toBe(false);
 });
+
+const assignmentBindingPatterns = [
+    { bindingString: "{x: obj.prop}", objectString: "{x: true}", returnVariable: "obj.prop" },
+    {
+        bindingString: "{x: obj.prop = true}",
+        objectString: "{x: undefined}",
+        returnVariable: "obj.prop",
+    },
+    { bindingString: "[{x: obj.prop}]", objectString: "[{x: true}]", returnVariable: "obj.prop" },
+    {
+        bindingString: "{obj: {prop: obj.prop}}",
+        objectString: "{obj: {prop: true}}",
+        returnVariable: "obj.prop",
+    },
+    { bindingString: "{x = true}", objectString: "{}", returnVariable: "x" },
+    {
+        bindingString: "{x: {[2 + 1]: y}}",
+        objectString: "{x: {[2 + 1]: true}}",
+        returnVariable: "y",
+    },
+];
+
+test.each([...assignmentBindingPatterns, ...testCases])(
+    "Binding pattern expressions (%p)",
+    ({ bindingString, objectString, returnVariable }) => {
+        const result = util.transpileAndExecute(`
+            let x, y, z, foo, bar, obj: { prop: boolean };
+            obj = { prop: false };
+            (${bindingString} = ${objectString})
+            return ${returnVariable};
+        `);
+        expect(result).toBe(true);
+    }
+);
+
+test.each([...assignmentBindingPatterns, ...testCases])(
+    "Binding patterns expressions pass conditional checks (%p)",
+    ({ bindingString, objectString, returnVariable }) => {
+        const result = util.transpileAndExecute(`
+            let x, y, z, foo, bar, obj: { prop: boolean };
+            obj = { prop: false };
+            if (${bindingString} = ${objectString}) {
+                return ${returnVariable};
+            }
+        `);
+        expect(result).toBe(true);
+    }
+);
+
+test.each([
+    { bindingString: "{ x: x.prop = true } = {}", returnValue: "x.prop", expectedResult: true },
+    {
+        bindingString: "{ x: x.prop = true } = {}",
+        returnValue: "typeof y === 'object'",
+        expectedResult: true,
+    },
+])("Binding pattern assignment pass-through (%p)", ({ bindingString, returnValue, expectedResult }) => {
```

Comment:
Similar to the above, it would be good to test the case where the object literal has a value for `x`.